Update cnos.py

istaicu opened the pull request:
Added "no logging terminal"  after log in.

##### SUMMARY
The problem was that, when trying to do a configuration on a Lenovo switch, in some cases, a timeout would occur with the following trace:

020-03-12 10:37:26,564 p=5491 u=root n=ansible | Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/ansible/utils/jsonrpc.py", line 45, in handle_request
    result = rpc_method(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/cliconf/__init__.py", line 42, in wrapped
    return func(self, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/cliconf/cnos.py", line 102, in edit_config
    results.append(self.send_command(**line))
  File "/usr/lib/python2.7/site-packages/ansible/plugins/cliconf/__init__.py", line 127, in send_command
    resp = self._connection.send(**kwargs)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/__init__.py", line 35, in wrapped
    return func(self, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/network_cli.py", line 580, in send
    response = self.receive(command, prompt, answer, newline, prompt_retry_check, check_all)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/network_cli.py", line 529, in receive
    data = self._ssh_shell.recv(256)
  File "/usr/lib/python2.7/site-packages/paramiko/channel.py", line 665, in recv
    out = self.in_buffer.read(nbytes, self.timeout)
  File "/usr/lib/python2.7/site-packages/paramiko/buffered_pipe.py", line 156, in read
    self._cv.wait(timeout)
  File "/usr/lib64/python2.7/threading.py", line 362, in wait
    _sleep(delay)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/network_cli.py", line 596, in _handle_command_timeout
    raise AnsibleConnectionFailure(msg)

After investigating, found out that the timeout occurred because some logs appear in the ssh terminal session in some cases, and the function that waits for the prompt, does not recognize the prompt in the response message, unless it is found at the end of the response. But in this case, the prompt is mixed up with the log message text. (This happens, for example, when trying to raise an interface (no shut command on an interface). When this command is issued, a log message immediately appears, showing the new state of the interface.)
And because of this timeout message, a chain reaction would occur and the switch could no longer be configured (timeout => switch stuck in the current submenu, and when trying to issue other cli commands, errors appear due to incorrect prompt).

The solution is to disable logging when ansible connects to the switch. This does not affect other connections in any way because the command takes effect only for the current session. Also, you would not be able to see those log messages anyway since there is no way to access that ansible ssh session, rendering them unnecessary.

 ##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
plugins/terminal/cnos.py

##### ADDITIONAL INFORMATION
Tested with the changes I propose and the timeout does not occur anymore. The switch configuration can be completely executed.

gundalow commented:
Can you please rebase your branch, that should fix the CI failure.

felixfontein commented:
Also, please add a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to).

istaicu commented:
Hi,

I've done a rebase and fixed a minor syntax error. Now the shippable tests pass. I've also added a changelog fragment.

Thank you,
Ioana

gundalow commented:
@amuraleedhar @dkasberg as maintainers of the CNOS Ansible modules could you please take a look at this?
Thanks in advance

amuraleedhar commented:
I am no longer working with Lenovo. Lenovo has closed down its networking division. Dave Kasberg is long retired and is leading an intersting life outside software. I am currently employed with Cisco and wont be able to take care of this. I still admire and will be proud to be part of Ansible family.
@amuraleedhar

Sent from Yahoo Mail on Android 
 
  On Thu, Jul 9, 2020 at 22:59, John R Barker<notifications@github.com> wrote:   


@amuraleedhar @dkasberg as maintainers of the CNOS Ansible modules could you please take a look at this?
Thanks in advance

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.

istaicu commented:
@gundalow 
Hi,
Kind reminder about this pull request. Although Anil is no longer with us and Lenovo networking division is no longer an active project, is will be quite a few years until it will be completely shut down.  Meanwhile, we still have to fix any issue we find. I am now in charge of the ansible part, and this is the change we need to make. Can you please integrate it into the main branch? 

Thanks,
Ioana

felixfontein commented on changelogs/fragments/terminal_plugin_cnos_update.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - Prevent timeout connection failure by adding "no logging terminal" after log in, in terminal plugin for cnos.
```
```suggestion
  - cnos terminal plugin - prevent timeout connection failure by adding "no logging terminal" after log in (https://github.com/ansible-collections/community.network/pull/16).
```

gundalow commented:
@amuraleedhar Anil, it's been a pleasure to work with you over the years. I wish you all the success at Cisco. FYI your GitHub provile still mentions Lenovo

@dkasberg Not sure if you'll see this, it's also been great working with you and I wish you luck in your future adventures.

@istaicu Thank you for keeping an eye on Lenovo networking. In https://github.com/ansible-collections/community.network/pull/100 I've set you as a maintainer of these modules, so we know we can just hit merge on your PRs. Please do shout out if there is anything we can do to assist you, it's great to see that even though this isn't an active project you are still helping keep Ansible support going.

istaicu commented:
@gundalow Thank you very much for your support. I'll keep bugging you if there are things I don't know. Thanks.

istaicu commented:
@gundalow Can you please make Mihai Broc ( @mihaibroc ) a maintainer? 
Thanks

gundalow commented:
> @gundalow Can you please make Mihai Broc ( @mihaibroc ) a maintainer too? To have a backup.
> Thanks

Done. Welcome to the Ansible community @mihaibroc
